pkg/client/record/event.go

```diff
@@ -266,13 +275,22 @@ func (recorder *recorderImpl) generateEvent(object runtime.Object, timestamp unv
 	event := recorder.makeEvent(ref, eventtype, reason, message)
 	event.Source = recorder.source
 
+	recorder.wg.Add(2)
 	go func() {
 		// NOTE: events should be a non-blocking operation
+		// NOTE: however, if you call this from short-living programs, you should
+		// call Flush() to make sure all events have been recorded.
 		defer utilruntime.HandleCrash()
+		defer recorder.wg.Done()
 		recorder.Action(watch.Added, event)
 	}()
 }
 
+// Flush will make sure all events are successfully recorded.
+func (recorder *recorderImpl) Flush() {
+	recorder.wg.Wait()
```

Comment:
Given a `DropIfChannelFull` and the wait group you're using, you could end up never returning because the mux never delivered your event to a watcher because its personal queue was full.